design-proposals/shadow-node.md

```diff
@@ -0,0 +1,86 @@
+
+2. Parts of Heartbeat
+    1. [labelNodeUnschedulable](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L96) is similarly as markNodeAsUnschedulable used only when virt-handler is stopping.
+
+    2. As part of [do](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L119) we do 2 calls:
+     [Get](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) - irrelevant as this is read access.
+    [Patch](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) - updating NodeSchedulable, CPUManager, KSMEnabledLabel labels and VirtHandlerHeartbeat KSMHandlerManagedAnnotation annotations once per minute.
+
+3. As part of node labeller
+    1. [run](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L195) - irrelevant as this is read access
+
+    2. [patchNode](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L254) - node labeller patching node with labels for various informations which are mostly static in nature:  cpu model, hyperv features, cpu timer, sev/-es, RT...
+This is triggered both by interval (3 min) and on changes of cluster config. 
+
+## Shadow node
+
+The proposed solution is to introduce new CRD that will be shadowing a node "ShadowNode". The virt-handler will be writting to ShadowNode and virt-controller will have a responsibility to sync allowed information to Node.
+
+## API Examples
+```yaml
+apiVersion: <apiversion>
+kind: ShadowNode
+metadata:
+  name: <Equivalent to existing name>
+  // Labels and Annotations only
+spec: {} //Empty, allows futher extension
+status: {} //Empty, allows futher extension
+```
+
+## Scalability
+
+There are few aspects to consider. 
+
+1.  #shadowNodes will be equivalent #nodes, negligible space overhead
+
+2. #writes could double. Here it is importent to look at the usage and sort each case to 2 categories (low and high frequent write). The first usage is clearly low frequent as long as virt-handler operates as designed.
+The second usage consist of two cases which might seem different but in fact these are same most of the time because NodeSchedulable, CPUManager, KSMEnabledLabel labels and KSMHandlerManagedAnnotation annotation is mostly static. What is left is VirtHandlerHeartbeat that is not necessary to sync on a Node (requires re-working node controller).
+The third case is propagating labels that are also mostly static.
+With all the above doubling #writes is unlikely. 
+
+
+## Update/Rollback Compatibility
+
+Virt-handler is going to continue writting to Node object, therefore update should be without compatibility issues.
```

Comment:
When upgrade rolls out, virt-controllers are upgraded first then the daemonsets followed by virt-api. There will be a timeperiod where new virt-controllers (with ShadowNode code) will work with old virt-handlers and old-apiserver without ShadowNode API. Can we mention here clear steps how:
1. new virt-controllers will work with older virt-handlers and olde virt-api
2. new virt-controllers and newer virt-handlers will work with older virt-api

Reply:
Hey @alaypatel07, I'll be taking over addressing the issue this proposal is trying to mitigate. I plan to take the same course this proposal is choosing. I will answer your question in the next reply, but if you think I should re-admit the proposal with the answers addressed in them - please tell me and I will do so.

Reply:
In regard to your question, I think the upgrade order is a bit different. Please keep me honest, but according to [this](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-operator/resource/apply/reconcile.go#L526-L641) - virt-operator upgrade order is like this:
```
CRDs --> RBACs --> Daemonsets (=virt-handler) --> Controller Deployments (=virt-controller)--> API Deployments (=virt-API)
```
Also note that as mentioned above, the virt-handler will still keep trying to patch the node alongside the shadowNode, and at some point during the upgrade, it would start being blocked due to the removal of the patch ability. In order to mitigate this, the node patch will start ignoring RBAC "forbidden" error.

Having said the above, this is a specific mention to each upgrade scenario - before and after the upgrade itself:
 
 During upgrade rollout:
* New CRD; old RBAC; old virt-handler; old virt-controller; old virt-API:
  * Old RBAC: no upgrade issue
  * Old virt-handler: no upgrade issue
  * Old virt-controllers: no upgrade issue
  * Old Virt-API: no upgrade issue
* New CRD, new RBAC; old virt-handler; old virt-controller; old virt-API:
  * Old virt-handler: during upgrade, the old RBACs are kept as backup until upgrade is done. Hence, patching node should not fail so there is no upgrade issue.
  * Old virt-controllers: no upgrade issue, virt controller RBAC is not changed due to this solution.
  * Old Virt-API: no upgrade issue.
* New CRD; new RBAC; new virt-handler; old virt-controller; old virt-API:
  * new virt-handler: shadowNode is written to but it is not synced to the node yet. However, virt-handler keeps patching the node directly, so there are no upgrade issue.
  * [same as before] Old controllers: no upgrade issue, virt controller RBAC is not changed. 
[same as before] Old Virt-API: no upgrade issue.
* New CRD; new RBAC; new virt-handler; new virt-controller; old virt-API:
  * New virt-handler: no upgrade issue. virt-handler patches both node and shadow node. 
  * New controllers: will start getting shadowNode requests once they are created. no upgrade issue.
  * [same as before] Old Virt-API: no upgrade issue.

After upgrade rollout:
virt-handler will keep trying (and failing) to patch the node since the backup RBACs are now removed. Virt-handler will ignore these cases by ignoring RBAC "forbidden" errors. This behavior should be kept until we no longer support upgrades from the current release.

Reply:
My suggestion, if possible, would be the following:

 * Introduce a new status field in the KubeVirt CR
 * Let virt-handlers continue patching the node as well
 * After virt-handlers and virt-controller are rolled over, set in the kubevirt CR a status field indicating that shadow nodes are now the way to go
  * virt-handler would pick it up right away over our config logic
  * finally remove the RBAC permissions.

Reply:
There may be some neglectible overlap between virt-handlers picking up the config change and the rbac removal then, but that should worst case just result in a few rbac permission errors which we should then _not_ mask.

Reply:
Hey @rmohr, thank you for your suggestion. I understand from what you say that you are concerned with masking the error. Am I right?
The way I see it, our options are:
1. (your suggestion) Adding a new status field. virt-operator will set it to true once [infrastructureRolledOver](https://github.com/kubevirt/kubevirt/blob/4179837bd7c20d991ae7a65d021b1f9a5480e1ae/pkg/virt-operator/resource/apply/reconcile.go#L519) is true. Everytime virt-handler wants to patch the node, it will check first if that status is true. 

Pros: 
    - avoid ignoring an error.
Cons: 
    - adding a field that will stay forever just for our little short-lived cause.
    - this has a potential for pesky corners issues (What if the upgrade failed somehow after the status was set? etc..)
    - there is a time gap until the config gets updated to each virt-handler via the informer - as you mentioned.

2. Checking before patching the node. Using the [SelfSubjectAccessReview](https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/self-subject-access-review-v1/#SelfSubjectAccessReview) API in order to check if it's allowed to patch the node before doing it.

Pros: 
    - avoids ignoring an error.
    - robust and stateless. We don't care when the upgrade rollout is done. Instead we check the API server directly. 
Cons: 
    - [minor] need to add RBAC to virt-handler ("authorization.k8s.io", "SelfSubjectAccessReview", "create", "get", "delete").
    - there is the small window between us checking "am I allowed to patch node" and actually doing the patch, where things could change and end up with the forbidden error anyway.

3. (The current proposal) Try patching, ignoring if failing. no need to say more. 

Pros: 
    - robust and stateless. We don't care when the upgrade rollout is done. Instead we try and maybe fail.
Cons: 
    - masking an error.
    - hard to predict if will cause issues in the future (forward compatibility)

4. (optimizing the current proposal) Upon failing once to patch on "forbidden" error - stop patching the node. This is what Lubo did on the high frequency hearbeat labelling in his original [PR](https://github.com/kubevirt/kubevirt/pull/10131/commits/ab963f50144b16784d093cd04ac1137d0832bff5#diff-dd05fa462e124061de87457f1df303157a338be8a163ea85e91aeffa6fa40721R170)

I personally tend towards the SelfSubjectAccessReview option, or masking the error like originally suggestion, since the error we are ignoring is pretty specific, and I don't see how it can conflict in the future.

Reply:
Hm, feels like SelfSubjectAccessReview way would just be another (expensive) way to just mask errors again (so we again just hope that it is correct that the permissions are gone). An extra status field would  from that perspective be much cheaper ...

Whenever we talk about masking we actually don't know if it is expected and I personally prefer to not add things where we just hope that things are right and swallow something.

Reply:
@rmohr, I respectfully disagree. Using SelfSubjectAccessReview isn't just masking errors; it's a direct and intentional way of asking the RBAC server, "Can I patch the node?" rather than just hoping things are right. It cheaper than patching the actual node then failing, and ensures actions are only taken when permitted, adding a layer of security. 

I believe relying on the upgrade mechanism to signal a status field once ready could prove to be more costly in the long run, considering the potential corner scenarios mentioned earlier.

Reply:
> @rmohr, I respectfully disagree. Using SelfSubjectAccessReview isn't just masking errors; it's a direct and intentional way of asking the RBAC server, "Can I patch the node?" rather than just hoping things are right.

I am not sure. I tried to express a knowledge gap which we try to overcome with checking for forbidden error or the SelfSubjectAccessReview. In both cases we don't  know if it is intentional and correct that we don't have the permission. With a clear config switch we know: Now I should do both, and from now on I should do one. Having a very brief period regarding to eventual consistency is fine where maybe one or two update requests would fail (like a low amount of conflict REST calls is fine too).

The information gap is what I consider "masking". Therefore the reviews add up quite a big front of additional uncertainties without making anything better (actually imo much more complex).

> It cheaper than patching the actual node then failing, and ensures actions are only taken when permitted, adding a layer of security.

I am not sure what "cheap" means here, and giving additional permissions which may leak info which we did not have access to before opens security questions which I don't even want to have to think about.

Reply:
@alaypatel07  I updated the upgrade scenarios on the proposal's body as you suggested.

Reply:
> @rmohr I am still against using the upgrade as a sign to know if we can stop patching the node. The upgrade process is complex and state-full, and relying on it may introduce corner cases. In a way, Lubo's "try and ignore if fails on specific error" is much more stateless and clean.
Can we agree to continue with his approach?

It should basically be a check like this:


```
if freshInstall && supportsShadowNodes()  {
  kubevirtCR.Status.UseShadowNode = True
}

if fully installed && !switchedToShadowNode && supportsShadowNodes() {
  kubevirtCR.Status.UseShadowNode = True
} else {
  kubevirtCR.Status.UseShadowNode = False
}
```

where `supportsShadowNodes()` can basically just check if the install strategy of the current version includes the shadow node CR, to ensure that the version we are upgrading downgrading to, has them.